tool/canonical/scorecard.dart

```diff
@@ -0,0 +1,403 @@
+          sb.write(' ${bugReferences.join(", ")} |');
+          break;
+      }
+    }
+    return sb.toString();
+  }
+
+  @override
+  String toString() => '$name$_ruleSets${hasFix ? " $bulb" : ""}';
+}
+
+class ScoreCard {
+  List<LintScore> scores = <LintScore>[];
+
+  int get lintCount => scores.length;
+
+  void add(LintScore score) {
+    scores.add(score);
+  }
+
+  String asMarkdown(List<Detail> details,
+      {int Function(LintScore s1, LintScore s2) sorter}) {
+    // Header.
+    var sb = StringBuffer();
+    details.forEach((detail) => sb.write('| ${detail.name} '));
+    sb.write('|\n');
+    details.forEach((detail) => sb.write(detail.header.markdown));
+    sb.write(' |\n');
+
+    if (sorter != null) {
+      scores.sort(sorter);
+    }
+
+    // Body.
+    forEach((lint) => sb.write('${lint.toMarkdown(details)}\n'));
+    return sb.toString();
+  }
+
+  void forEach(void Function(LintScore element) f) {
+    scores.forEach(f);
+  }
+
+  void removeWhere(bool Function(LintScore element) test) {
+    scores.removeWhere(test);
+  }
+
+  static Future<ScoreCard> calculate() async {
+    var lintsWithFixes = await _getLintsWithFixes();
+    var lintsWithBulkFixes = await _getLintsWithBulkFixes();
+    var lintsWithAssists = await _getLintsWithAssists();
+
+    // var issues = await _getIssues();
+    // var bugs = issues.where(_isBug).toList();
+    var bugs = <Issue>[];
+
+    var scoreRuleset = _readScoreLints();
+    var recommendRuleset = _readRecommendLints();
+
+    var scorecard = ScoreCard();
+    for (var lint in registeredLints) {
+      var ruleSets = <String>[];
+      if (scoreRuleset.contains(lint.name)) {
+        ruleSets.add('score');
+      }
+      if (recommendRuleset.contains(lint.name)) {
+        ruleSets.add('recommend');
+      }
+
+      if (ruleSets.isEmpty) {
+        continue;
+      }
+
+      var bugReferences = <String>[];
+      for (var bug in bugs) {
+        if (bug.title.contains(lint.name)) {
+          bugReferences.add('#${bug.number.toString()}');
+        }
+      }
+
+      var lintName = lint.name;
+      scorecard.add(LintScore(
+        name: lintName,
+        hasFix: lintsWithFixes.contains(lintName) ||
+            lintsWithAssists.contains(lintName),
+        hasBulkFix: lintsWithBulkFixes.contains(lintName),
+        maturity: lint.maturity.name,
+        ruleSets: ruleSets,
+        bugReferences: bugReferences,
+      ));
+    }
+
+    return scorecard;
+  }
+
+  // static Future<List<Issue>> _getIssues() async {
+  //   var github = GitHub();
+  //   var slug = RepositorySlug('dart-lang', 'linter');
+  //   try {
+  //     return github.issues.listByRepo(slug).toList();
+  //   } on Exception catch (e) {
+  //     print('exception caught fetching github issues');
+  //     print(e);
+  //     print('(defaulting to an empty list)');
+  //     return Future.value(<Issue>[]);
+  //   }
+  // }
+
+  static Future<List<String>> _getLintsWithAssists() async {
```

Comment:
I don't understand what we're trying to do with this information, but the extra data on `AssistKind` isn't being maintained (and I'll have a CL shortly to remove it). We're now using `FixProcessor.lintProducerMap` and depending on the fact that assists and fixes that do the same thing share the same producer in order to determine whether an assist has a fix associated with it.

Reply:
Ah!  OK, right. Much better.  Will update.

Reply:
The rub here (and below) is that `FixProcessor` is in server and we don't have a dependency on server (and can't since it isn't published).

Reply:
Oh, right. Sigh. Mono-repos are so much more convenient sometimes.

Reply:
That said, I still don't understand why we want this data at all. Every lint mentioned in an `AssistKind`'s list is already also mentioned in the `FixProcessor`, so I don't see how this list provides value. Does something break if we were to stop calling this method?

Reply:
Ah, I think this can be simplified.  The code I was referencing dates back to a time when I *think* there were assists w/o corresponding fixes.  Updating.  Thanks!